teerex/src/lib.rs

```diff
@@ -237,13 +245,40 @@ pub mod pallet {
 				sender_enclave.mr_enclave.encode() == shard_id.encode(),
 				<Error<T>>::WrongMrenclaveForShard
 			);
-			<WorkerForShard<T>>::insert(shard_id, sender_index);
-			log::debug!(
-				"Proposed sidechain block confirmed with shard {:?}, block hash {:?}",
-				shard_id,
-				block_hash
-			);
-			Self::deposit_event(Event::ProposedSidechainBlock(sender, block_hash));
+
+			let mut latest_block_header = <LatestSidechainBlockHeader<T>>::get();
+			let mut latest_block_number = latest_block_header.block_number;
+			let block_number = block_header.block_number;
+			log::error!("Got new block: {:?}", block_number);
+			if latest_block_number + 1 == block_number {
+				latest_block_header = Self::check_block_and_get_latest_header(
+					shard_id,
+					block_header,
+					&sender,
+					sender_index,
+					latest_block_header,
+				);
+			} else if latest_block_number + 1 < block_number {
```

Comment:
What happens if `block_number < latest_block_number + 1` ? It seems we then still set `latest_block_number = latest_block_header.block_number`, which is probably not correct then? In that case we should just discard the header and do nothing, I believe?

Reply:
Yes, that is what I understood, that we discard headers with the same numer that are already imported.

Reply:
Okay, but that's not what happens in the code as I see it here?

Reply:
yes, we chan skip the rest of the method for  latest_block_number >= block_number

Reply:
Additionally, I want to have a safety-param here: 

```
latest_block_number + 1 < block_number + EARLY_BLOCK_PROPOSAL_LENIENCE
```

To prevent the arrival of blocks that are far too early. If this check fails, an error should be returned.

This safety param should be introduced as a runtime-constant.

Reply:
I honestly don't like the approach of "doing nothing". This is hell for every developer trying to debug. Can't you return an error instead?

Reply:
But it is not an error? When we encounter such a header, we need to discard/ignore it. This is not an error case, it's part of the regular operation of this function. In addition to testing (which demonstrates the mechanism), we could add a `debug!()` or `trace!()` log.

Reply:
Let's schedule a meeting. :)

Reply:
Let's return an error

Reply:
We decided to return the errors
 * BlockNumberTooHigh in case the blocknumber is lenience above the latest blocknumber and
 * OutdatedBlockNumber in case the blocknumber is below the latest blocknumber.